arm: sample apps and build support for PSA TF-M and Zephyr

karl-zh opened the pull request:
This PR enables CI checks to pass for TF-M, and introduces build support to automatically pull in the external TF-M repository and all dependencies, and perform a build of TF-M as part of a standard Zephyr sample app build procedure. It is based on TF-M 1.0-rc2.

Commit:
arm: mps_an521_nonsecure: Add TFM BL2 support

PSA level 1 requires secure boot. TF-M BL2 is the official
secure boot loader. It needs a BL2_HEADER_SIZE offset.

Align nonsecure address with TF-M's NS slot while TF-M BL2 enabled.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

microbuilder commented on samples/tfm_integration/tfm_psa_arch_tests/src/main.c:
```diff
@@ -0,0 +1,37 @@
+	.stack_size = sizeof(test_app_stack),
+};
+
+static osStatus_t     status;
```
```
At top level:
../src/main.c:24:23: warning: 'status' defined but not used [-Wunused-variable]
 static osStatus_t     status;
```

microbuilder commented on samples/tfm_integration/tfm_psa_arch_tests/src/main.c:
```diff
@@ -0,0 +1,37 @@
+void main(void)
+{
+	/* Initialize the TFM NS interface */
+	tfm_ns_interface_init();
```
```
../src/main.c:31:2: warning: implicit declaration of function 'tfm_ns_interface_init' [-Wimplicit-function-declaration]
  tfm_ns_interface_init();
  ^~~~~~~~~~~~~~~~~~~~~
```

microbuilder commented on samples/tfm_integration/tfm_psa_arch_tests/src/main.c:
```diff
@@ -0,0 +1,37 @@
+	/* Initialize the TFM NS interface */
+	tfm_ns_interface_init();
+
+	thread_func = val_entry;
```
```
../src/main.c:33:14: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
  thread_func = val_entry;
```

microbuilder reviewed (changes requested):
Looks good other than some minor cleanup issues that should be easy to resolve.

I've run both of these samples and they work fine in QEMU with the following script:

```
#!/bin/bash
#-----------------------------------------------------------------------------
# Copyright 2019 Linaro Limited
# SPDX-License-Identifier: Apache-2.0
#
# Merges a secure TF-M library that enables IPC communications with a
# non-secure Zephyr application.
#
# This files is a workaround since we can't currently insert a real post-build
# step in the Zephyr build system, so the zephyr and TF-M binaries need to
# be manually merged pending changes to the build system.
#-----------------------------------------------------------------------------

# Halt on errors
set -e
QEMUBOARD="mps2-an521"

# Merge bootloader and TFM image into single binary
srec_cat build/mcuboot.bin -Binary build/tfm_sign.bin -Binary -offset \
0x80000 -o build/tfm_qemu.bin -Binary

# Convert .bin to .elf with an appropriate offset
srec_cat build/tfm_qemu.bin -binary -offset 0x10000000 -o build/tfm_qemu.hex \
-intel --line-length=44

# Start qemu
echo "Starting QEMU (CTRL+C to quit)"
qemu-system-arm -M ${QEMUBOARD} -device loader,file=build/tfm_qemu.hex \
-serial stdio
```

I also was able to add the PSA Certified Level 1 sample app on top of this PR, and will likely rebase #18185 on top of this, or close that PR and include the PSA sample app here as an all-in-one TF-M PR with all three sample apps included.

karl-zh commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,43 @@
+		COMMAND ${PYTHON_EXECUTABLE} ${MCUBOOT_DIR}/scripts/imgtool.py
+		ARGS sign
+			 --layout ${PREPROCESSED_FILE}
+			 -k ${MCUBOOT_DIR}/root-rsa-3072.pem
```
We now based on a commit for psa level 1 of Zephyr. I think keep update with TF-M is endless at the moment.

karl-zh commented on ext/tfm/Kconfig:
```diff
@@ -0,0 +1,19 @@
+	depends on PSA_ARCH_API_TEST
+	string "PSA arch API test suite"
+	help
+	  This indicates which test suite will be action.
```
Fixed in a new commit. Thanks.

carlescufi reviewed (changes requested):
This should be a west project and a zephyr module. Please open an issue to create a TF-M repo under the zephyrproject-rtos and then be able to point `west.yml` to it. Please follow the instructions here:

https://docs.zephyrproject.org/latest/guides/modules.html

karl-zh commented:
> This should be a west project and a zephyr module. Please open an issue to create a TF-M repo under the zephyrproject-rtos and then be able to point `west.yml` to it. Please follow the instructions here:
> 
> https://docs.zephyrproject.org/latest/guides/modules.html

Hi,
The TF-M project and it's related repos are standalone projects. And the TF-M is not a module or a part of Zephyr. TF-M run in secure execution and Zephyr run in nonsecure execution, the compile seperately to different binaries. May be this requires more discussions or ideas on this point for further. Anyway, created an issue for new module request.

Thanks

Commit:
arm: musca_b1: Extend memory space in nonsecure

This is for PSA arch test to run on Musca B1. The test cases need more
than 100k RAM.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

karl-zh commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -27,7 +27,7 @@ else()
 
 # Select a known trusted-firmware-m commit
-set(TFM_COMMIT_HASH "5358c8c4cf385cde6091adc74dce53fa3f0c2ce4")
+set(TFM_COMMIT_HASH "4117e0353c55dce739f550515c5b007411142ba7")
```
This could specific to a tag I think. TF-Mv1.0-RC2
I may update this in previous patch.

karl-zh reviewed (commented):
What is this commit for ?
"Merge github.com:zephyrproject-rtos/zephyr into psa_arch_tests_for_tf…"

karl-zh commented:
@microbuilder I am working on Musca B1 to make it reproduce psa arch tests pass. The Zephyr is 2.0.0. Output like below:

`[Sec Thread] Secure image initializing!
***** Booting Zephyr OS build zephyr-v2.0.0-1393-gec3f603b8a7a *****
TF-M PSA Arch Tests with Zephyr on musca_b1_nonsecure

***** PSA Architecture Test Suite - Version 0.9 *****

Running.. Crypto Suite
******************************************
....

************ Crypto Suite Report **********
TOTAL TESTS     : 44
TOTAL PASSED    : 43
TOTAL SIM ERROR : 0
TOTAL FAILED    : 0
TOTAL SKIPPED   : 1
******************************************

Entering standby.. 

`

microbuilder commented:
> What is this commit for ?
> "Merge github.com:zephyrproject-rtos/zephyr into psa_arch_tests_for_tf…"

This is a consequence of updating the PR against master, pulling recent changes into the PR ... I can remove this and most other superfluous commit messages before the final PR review to clean this up.

microbuilder commented:
> This should be a west project and a zephyr module. Please open an issue to create a TF-M repo under the zephyrproject-rtos and then be able to point `west.yml` to it. Please follow the instructions here:
> 
> https://docs.zephyrproject.org/latest/guides/modules.html

Hi Carles ... I understand the idea behind the module request, but I would argue in this case that this is specifically a **sample** with some BSP fixes thrown on top, and not a module as such since it doesn't add TFM in the PR itself. The cmake file does do some work to pull TF-M in as an external project as part of the normal Zephyr build process, build it with appropriate options, and link against that for the secure processing environment, but the core code here is exclusively a sample application for Zephyr, which doesn't seem inappropriate to me to include in the samples folder(?).

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,230 @@
+On MPS2+ AN521:
+===============
+
+1. Build Zephyr with a non-secure configuration
```
The "using west" and "using cmake" instructions are part of step 1, so should be indented:

```
1. Build Zephyr with a non-secure configuration
   (``-DBOARD=mps2_an521_nonsecure``). This will also build TF-M in the
   ``ext/tfm`` folder if necessary.

   Using ``west``:

   .. code-block:: bash

      cd <ZEPHYR_ROOT>/samples/tfm_integration/psa_level_1/
      west build -p -b mps2_an521_nonsecure ./

   Using ``cmake``:

   .. code-block:: bash

      cd <ZEPHYR_ROOT>/samples/tfm_integration/psa_level_1/
      mkdir build
      cd build
      cmake -GNinja -DBOARD=mps2_an521_nonsecure ..
      ninja -v
```

dbkinder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -0,0 +1,230 @@
+   (``-DBOARD=mps2_an521_nonsecure``). This will also build TF-M in the
+   ``ext/tfm`` folder if necessary.
+
+Using ``west``
```
same indenting issue here as noted above

dbkinder commented on samples/tfm_integration/tfm_integration.rst:
```diff
@@ -0,0 +1,10 @@
+.. _tfm_integration-samples:
+
+Various TFM Integration Samples
```
I'd delete ``Various`` from this title.

dbkinder commented on samples/tfm_integration/tfm_ipc/README.rst:
```diff
@@ -0,0 +1,84 @@
+On MPS2+ AN521:
+===============
+
+1. Build Zephyr with a non-secure configuration (``-DBOARD=mps2_an521_nonsecure``).
```
The code block below should be indented to be part of step 1:

```
1. Build Zephyr with a non-secure configuration (``-DBOARD=mps2_an521_nonsecure``).

   .. code-block:: bash

      cd $ZEPHYR_ROOT/samples/tfm_integration/tfm_ipc/
      mkdir build
      cd build
      cmake -GNinja -DBOARD=mps2_an521_nonsecure ..
      ninja -v
```

dbkinder commented on samples/tfm_integration/tfm_ipc/README.rst:
```diff
@@ -0,0 +1,84 @@
+3. Open ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt``.
+4. Update the ``AN521/images.txt`` file as follows:
+
+.. code-block:: bash
```
same indenting issue noted above for this code-block too

dbkinder commented on samples/tfm_integration/tfm_psa_arch_tests/README.rst:
```diff
@@ -0,0 +1,127 @@
+to run in a secure configuration with Zephyr itself in a non-secure configuration.
+The sample prints test info to the console as a multi-thread application.
+
+The psa-arch-tests provides differents test suits, this example supports
```
I think you're saying you can specify one of these specific test suites. If so, I'd rewrite this as:
```
The psa-arch-tests provides different test suite configurations:
``crypto``, ``protected_storage``, or ``initial_attestation``, as set in
the sample's ``prj.conf`` file, for example:

.. code-block:: bash

   CONFIG_API_TEST_SUIT="crypto"
```

(This also fixes some misspellings)

dbkinder commented on samples/tfm_integration/tfm_psa_arch_tests/README.rst:
```diff
@@ -0,0 +1,127 @@
+
+1. Build Zephyr with a non-secure configuration (``-DBOARD=mps2_an521_nonsecure``).
+
+.. code-block:: bash
```
This code-block needs to be indented (as shown previously)

dbkinder commented on samples/tfm_integration/tfm_psa_arch_tests/README.rst:
```diff
@@ -0,0 +1,127 @@
+3. Open ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt``.
+4. Update the ``AN521/images.txt`` file as follows:
+
+.. code-block:: bash
```
This code-block needs to be indented (as shown previously)

dbkinder commented on samples/tfm_integration/tfm_psa_arch_tests/README.rst:
```diff
@@ -0,0 +1,127 @@
+
+1. Build Zephyr with a non-secure configuration (``-DBOARD=v2m_musca_b1_nonsecure``).
+
+.. code-block:: bash
```
This code-block needs to be indented (as shown previously)

dbkinder commented on samples/tfm_integration/tfm_psa_arch_tests/README.rst:
```diff
@@ -0,0 +1,127 @@
+   cmake -GNinja -DBOARD=v2m_musca_b1_nonsecure ..
+   ninja -v
+
+2. The binaries will be sign and combine automatically into build folder (tfm_zephyr.hex).
```
Change to:
```
2. The binary file ``tfm_zephyr.hex`` will be signed and combined
   automatically into the build folder.
```

dbkinder commented on samples/tfm_integration/tfm_psa_arch_tests/README.rst:
```diff
@@ -0,0 +1,127 @@
+   ninja -v
+
+2. The binaries will be sign and combine automatically into build folder (tfm_zephyr.hex).
+3. Plunge in Musca B1 board, and press power on button.
```
Change to:
```
3. Connect the USB cable from your development system
   to the Musca B1 board, and press the power-on button.
```

dbkinder commented on samples/tfm_integration/tfm_psa_arch_tests/README.rst:
```diff
@@ -0,0 +1,127 @@
+
+2. The binaries will be sign and combine automatically into build folder (tfm_zephyr.hex).
+3. Plunge in Musca B1 board, and press power on button.
+4. Drag and drop tfm_zephyr.hex to root of the MUSCA_B(E:) drive.
```
Hmmm.  Sounds like you did this on a Windows system.  Can you write this to work on both windows and Linux?

dbkinder reviewed (approved):
Doc changes LGTM, thanks!

karl-zh commented:
Why combine all commit history into one in the PR?

microbuilder commented:
> Why combine all commit history into one in the PR?

It makes the commit easier to refer to in the future once it gets merged in, versus 15 related but overlapping commits.

karl-zh commented:
> > Why combine all commit history into one in the PR?
> 
> It makes the commit easier to refer to in the future once it gets merged in, versus 15 related but overlapping commits.

Hi Kevin,

I do not think so. The commits means history and the steps of how the feature enabled. All patches are trackable by git log. And also easier for reviewer. I do combine some patches because they just fix few words that in the same document. For other patches, I'd prefer keep them as original.

karl-zh commented on samples/tfm_integration/tfm_psa_arch_tests/sample.yaml:
```diff
@@ -5,4 +5,5 @@ sample:
 tests:
   sample.tfm_ipc:
     tags: introduction
-    platform_whitelist: mps2_an521_nonsecure v2m_musca_b1_nonsecure
```
I am not sure if the CI care about this build time strictly? I think this platform is good to build while CI check since this sample should support on Musca. Besides, the Zephyr PSA L1 was certified on Musca B1.

microbuilder replied:
Is there a specific benefit to running these long tests (750MB download for the TF-M repo and dependencies) twice? It seems like everything is tested already on the MPS2+ which is the defacto M33 qemu platform, and running musca as well would just add overhead without any clear benefit since the two devices and sample configs are functionally equivalent (the Musca B1 is just a silicon implementation of the code running on the FPGA on the MPS2+ w/AN521?

galak reviewed (changes requested):
So I think we should see about having TF-M as a west module and setting up a repo under the zephyr project that tracks the TF-M repo (and maybe combines all the components that TF-M needs as its own west manifest).

karl-zh commented:
> So I think we should see about having TF-M as a west module and setting up a repo under the zephyr project that tracks the TF-M repo (and maybe combines all the components that TF-M needs as its own west manifest).

Hi,
If put TFM and all dependencies as modules, is that means everyone will download them by west update, is it? And the modules seem using the zephyr CMake style to build each module, I think to do the same for TF-M project is a big effort. The TFM eventually not as a part of zephyr.bin, How do you think about the way we using now? To keep TF-M independent pull, and compile.

microbuilder commented:
> If put TFM and all dependencies as modules, is that means everyone will download them by west update, is it? And the modules seem using the zephyr CMake style to build each module, I think to do the same for TF-M project is a big effort. The TFM eventually not as a part of zephyr.bin, and all dependencies projects size is large, so implement all the work like it shows in this PR. To keep TF-M independent pull, and compile.

Yes, the modules would be installed via the usual `west update` command, and most likely this will require a fork of trusted-firmware-m.git into a new project like `zephyrproject-rtos/tfm` to host the TF-M code, and any support files similar to what happens with open-amp today: https://github.com/zephyrproject-rtos/open-amp

karl-zh commented:
> > If put TFM and all dependencies as modules, is that means everyone will download them by west update, is it? And the modules seem using the zephyr CMake style to build each module, I think to do the same for TF-M project is a big effort. The TFM eventually not as a part of zephyr.bin, and all dependencies projects size is large, so implement all the work like it shows in this PR. To keep TF-M independent pull, and compile.
> 
> Yes, the modules would be installed via the usual `west update` command, and most likely this will require a fork of trusted-firmware-m.git into a new project like `zephyrproject-rtos/tfm` to host the TF-M code, and any support files similar to what happens with open-amp today: https://github.com/zephyrproject-rtos/open-amp
> 
> This would have the advantage of making it clearer which commit of TF-M we are currently using, and we can also add the TF-M dependencies into this new repo (`mbed-crypto`, `CMSIS_5`, etc.) so that we no longer rely on an awkward `git clone` solution to build.

I know that, and was wrok on the openamp module. Since that script can be updated. https://github.com/zephyrproject-rtos/zephyr/commit/469ad5218be949ee421dc6d6355d5553fbd7ee5b

karl-zh commented:
> > So I think we should see about having TF-M as a west module and setting up a repo under the zephyr project that tracks the TF-M repo (and maybe combines all the components that TF-M needs as its own west manifest).
> 
> Hi,
> If put TFM and all dependencies as modules, is that means everyone will download them by west update, is it? And the modules seem using the zephyr CMake style to build each module, I think to do the same for TF-M project is a big effort. The TFM eventually not as a part of zephyr.bin, and all dependencies projects size is large, so implement all the work like it shows in this PR. To keep TF-M independent pull, and compile.

1. If put it in module, west update will have 730 MB larger for every one to include TF-M and it's dependent projects.
2. All Cmake files in TFM does not using zephyr_xxxx() now, and I do not think it is vaulable to rework all Cmake files on a specific TF-M version, TF-M evolves all the time.
3. tfm_s.bin is not a part in zephyr.bin, not sure if it is properly as a module.
Just my oponion on it, thanks.

galak commented:
> > > So I think we should see about having TF-M as a west module and setting up a repo under the zephyr project that tracks the TF-M repo (and maybe combines all the components that TF-M needs as its own west manifest).
> > 
> > 
> > Hi,
> > If put TFM and all dependencies as modules, is that means everyone will download them by west update, is it? And the modules seem using the zephyr CMake style to build each module, I think to do the same for TF-M project is a big effort. The TFM eventually not as a part of zephyr.bin, and all dependencies projects size is large, so implement all the work like it shows in this PR. To keep TF-M independent pull, and compile.
> 
> 1. If put it in module, west update will have 730 MB larger for every one to include TF-M and it's dependent projects.

we should be able to construct the tf-m module repo in such a way that its not that big, we don't need all of CMSIS for example.

> 2. All Cmake files in TFM does not using zephyr_xxxx() now, and I do not think it is vaulable to rework all Cmake files on a specific TF-M version, TF-M evolves all the time.
> 3. tfm_s.bin is not a part in zephyr.bin, not sure if it is properly as a module.
>    Just my oponion on it, thanks.

So I think we should look at a middle ground here to start with.  We should use west as the means of getting the TFM source, how we build it after that we can figure out.  I agree it doesn't make sense to introduce/change the TFM CMake files.  However we should be able to do something similar to openamp.

dbkinder commented on boards/arm/mps2_an521/doc/index.rst:
```diff
@@ -431,6 +424,8 @@ serial port:
 
    Hello World! mps2_an521
 
+.. note:: Please refer to the tfm_integration sample for more details about integrating with TF-M and multiple images scenario.
```
While it's polite, the project preference is not to use "please", so start the sentence with ``Refer to the...``

dbkinder commented on samples/tfm_integration/psa_level_1/docs/attestation.md:
```diff
@@ -0,0 +1,274 @@
+# Initial Attestation
```
Hmm.. Why isn't this a .rst file instead of a markdown file?  As is, it won't be included in our generated documentation. Check out ``pandoc`` as a quick way to convert ``.md`` to ``.rst``

karl-zh replied:
@microbuilder

dbkinder replied:
I can convert this to rst if you’d like.

karl-zh replied:
https://github.com/zephyrproject-rtos/zephyr/commit/0d3d2d85915889fffd9a077cbc8e5809a13a07c2
This commit can be amended afterward I think.

dbkinder commented on samples/tfm_integration/tfm_ipc/README.rst:
```diff
@@ -0,0 +1,84 @@
+      ninja -v
+
+2. Copy application binary files (mcuboot.bin and tfm_sign.bin) to ``<MPS2 device name>/SOFTWARE/``.
+3. Open ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt``.
```
These two steps should be combined to say:
```
#. Edit ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt`` as follows:
```

You can use ``#.`` for auto-numbering, instead of a hard numbered step.

dbkinder commented on samples/tfm_integration/tfm_psa_arch_tests/README.rst:
```diff
@@ -0,0 +1,128 @@
+      ninja -v
+
+2. Copy application binary files (mcuboot.bin and tfm_sign.bin) to ``<MPS2 device name>/SOFTWARE/``.
+3. Open ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt``.
```
Same comment here about combining into one step that says "Edit"

dbkinder reviewed (dismissed):
Please convert the markdown (.md) file to restructuredText (.rst) so it can be included in our generated documentation.

erwango commented:
@karl-zh, was this intentional to push a merge commit ?

microbuilder commented:
> This should be a west project and a zephyr module. Please open an issue to create a TF-M repo under the zephyrproject-rtos and then be able to point `west.yml` to it. Please follow the instructions here:
> 
> https://docs.zephyrproject.org/latest/guides/modules.html

This is now a module, and should resolve the change request on your part @carlescufi ?

carlescufi commented on west.yml:
```diff
@@ -118,6 +118,9 @@ manifest:
       path: tools/edtt
       revision: dd4dd502ef2fbeced6ef7faaba562a7ddca45632
+    - name: trusted-firmware-m
+      path: modules/tfm
```
Not sure if this should instead belong in `modules/tee/tfm`?
@galak @ioannisg @microbuilder

ioannisg replied:
Agreed, at least ARC might have their own TEE module, after all

carlescufi replied:
@galak @microbuilder does `modules/tee/tfm` make sense to you?

microbuilder replied:
`tee` seems fine to me, yes.

galak replied:
I moved it to that location.

microbuilder replied:
Fixed

galak commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,85 @@
 zephyr_library_include_directories(${ZEPHYR_BASE}/drivers)
+
+if (CONFIG_BUILD_WITH_TFM)
+	set (TFM_ROOT_PATH $ENV{ZEPHYR_BASE}/../modules/tee/tfm/trusted-firmware-m)
```
`TFM_ROOT_PATH` should be set in some common place not here

galak commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,85 @@
+
+if (CONFIG_BUILD_WITH_TFM)
+	set (TFM_ROOT_PATH $ENV{ZEPHYR_BASE}/../modules/tee/tfm/trusted-firmware-m)
+	set (MCUBOOT_DIR "${TFM_ROOT_PATH}/bl2/ext/mcuboot")
```
`MCUBOOT_DIR`, should be set in some common place and be renamed to `TFM_MCUBOOT_DIR`

galak reviewed (dismissed):
* The board integration has too much common stuff from TFM that should be set in a common place if you've enabled TFM.
* At some point once the changes look reasonable we should move some of the Cmake to the tfm module repo [suggesting doing that once we've got the structure of what's general TFM Cmake, board, and sample all set]

galak commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,85 @@
+		set(IMAGE_VERSION_NS 0.0.0+0)
+	endif()
+	#Find Python3 interpreter
+	find_package(PythonInterp 3)
```
Do we need this?  Is zephyr python setting not sufficient?

galak commented on boards/arm/v2m_musca_b1/v2m_musca_b1_nonsecure.dts:
```diff
@@ -42,14 +42,14 @@
 	};
 
-	flash0: flash@0a060400 {
+	flash0: flash@0a070400 {
```
The commit that changes this is for psa tests, do this in as `samples/tfm_integration/tfm_psa_arch_tests/boards/v2m_musca_b1.overlay`

microbuilder replied:
TF-M has also gotten larger with new features since the last 1.0rc2 commit, and I would propose just leaving the extra 64KB to allow room for future changes to the S image and bootloader?

galak commented on modules/Kconfig.tfm:
```diff
@@ -0,0 +1,17 @@
+config BUILD_WITH_TFM
+	bool "Build with TF-M for Secure Execution"
+
+config PSA_ARCH_API_TEST
```
Seems like PSA_ARCH_API_TEST and PSA_API_TEST_SUITE should just be if we're building the psa sample app.

Also wonder if those should move under tests/ instead of samples

microbuilder replied:
I agree these should be moved to the app level, but I don't think the app belongs under `tests` myself, and should probably stay as an individual sample app. They aren't really unit tests in the Zephyr sense, but rather something you need to run once to generate a report (for a specific board) that you send out to a test lab to have your board certified as being PSA FF API compliant. The sample app is provided as a convenience to generate the required report when certifying a board with Zephyr, not to test the API itself on a regular basis.

galak commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,92 @@
+project(NONE)
+
+# Make sure we are using a compatible board target.
+if((${BOARD} STREQUAL "mps2_an521_nonsecure"))
```
We shouldn't need BOARD stuff here, it should move into the sample/test app files.

galak commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,92 @@
+        message(FATAL_ERROR "${BOARD} unsupported. Please use \'${ZEPHYRBOARD}\'.")
+endif()
+
+if(${CONFIG_PSA_ARCH_API_TEST})
```
This should be with the sample/test psa api, not here.

galak commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,92 @@
+set(TFM_PSA_API_DIR "${TFM_BASE}/build/install/export/tee/tfm")
+
+# Assign the full git path to ${GIT_EXECUTABLE}.
+find_package(Git)
```
I don't think we need this anymore.

galak commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,92 @@
+endfunction()
+
+# Build psa-arch-tests before TF-M
+if(${CONFIG_PSA_ARCH_API_TEST})
```
I think this might be able to be reworked as a CMake `ExternalProject_Add` like the tfm ipc is doing.

microbuilder replied:
Changed to external project.

ioannisg commented on boards/arm/mps2_an521/doc/index.rst:
```diff
@@ -380,20 +380,13 @@ The process requires five steps:
 5. Concatenate the bootloader with the signed image blob.
 
-In order to build tfm please refer to `Trusted Firmware M Guide`_.
+To build tfm as secure image, refer to `Trusted Firmware M Guide`_.
```
Isn't this redundant @microbuilder ? I mean TF-M is by definition a secure image

microbuilder replied:
I guess I was just trying to indicate the distinction between TF-M as the source of the S binary, running in the secure processing environment, contrasted with Zephyr as the non-secure image, and the two images get merged as part of the build process. I can reword this if it doesn't seem intuitive as is, though.

microbuilder replied:
Updated in https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/9f39aeecfed25dffe1f225627f11cacb0424e362

carlescufi commented:
@tejlmand could you please take a look at the integration between the 2 CMake build systems?

microbuilder commented:
> @tejlmand could you please take a look at the integration between the 2 CMake build systems?

I havent been able to get `ninja` to do the TF-M build first, the way that `make` does. I believe the dependencies are being setup correctly, with the `add_dependencies` function:

```
ExternalProject_Add(
  tfm
  SOURCE_DIR ${TFM_BASE_DIR}
  CMAKE_ARGS -DPROJ_CONFIG=${TFM_BASE_DIR}/configs/ConfigRegressionIPC.cmake
             -DTARGET_PLATFORM=${TFMBOARD}
	     -DBL2=True
	     -DCOMPILER=GNUARM
	     #-DGNUARM_PREFIX=arm-zephyr-eabi
	     #-DGNUARM_PATH=${ZEPHYR_SDK_INSTALL_DIR}/arm-zephyr-eabi
	     -DGNUARM_PREFIX=arm-none-eabi
	     -DGNUARM_PATH=/usr/local/prog/gcc-arm-none-eabi-7-2018-q2-update
  BUILD_BYPRODUCTS ${BINARY_DIR}/tfm/veneers/s_veneers.o
)
add_dependencies(app tfm)
```

With ninja:

```
$ cd samples/tfm_integration/tfm_ipc/
$ mkdir build && cd build
$ cmake -GNinja -DBOARD=mps2_an521_nonsecure ..
$ ninja
ninja: error: 'tfm-prefix/src/tfm-build/install/export/tfm/veneers/s_veneers.o', needed by 'zephyr/zephyr_prebuilt.elf', missing and no known rule to make it
```

With make:

```
$ cmake -DBOARD=mps2_an521_nonsecure ..
$ make run
...
[TF-M build output]
...
[Sec Thread] Secure image initializing!
TF-M isolation level is: 1
Booting TFM v1.0
*** Booting Zephyr OS build v1.12.0-rc1-19023-g10461e3708e7  ***
The version of the PSA Framework API is 256.
The minor version is 1.
Connect success!
TFM service support minor version is 1.
```

tejlmand commented:
@microbuilder I changed a little in the CMakeLists.txt to get it running locally with `ninja`
You can try to fetch: https://github.com/karl-zh/zephyr/pull/1 and do `west update`.
This only corrects the dependency issues.

Note, there are still some build issues that @karl-zh  should look into.

tejlmand reviewed (changes requested):
Made a couple of small suggestions for #19985 just to get it to build.
Those are found here:
https://github.com/karl-zh/zephyr/pull/1
https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/3

The `s_veneers.o` file are identical but in two different location.
But it is important that the build system points to the same path.
That is, the path in 
```
target_link_libraries(${ZEPHYR_CURRENT_LIBRARY} PRIVATE
  ${TFM_EXPORT_PATH}/tfm/veneers/s_veneers.o
  )
```
must be the same that is desribed by `BUILD_BYPRODUCTS`.
And when moving `${TFM_EXPORT_PATH}` up to be used in `BUILD_BYPRODUCTS` then this variable must be available at earlier stage.

However, this does not fix everything, cause there still seem to be issues related to linking:
`cmake -GNinja -DBOARD=mps2_an521_nonsecure -Bbuild . `
works.

`ninja -C build -v`
fails with:
```
/opt/zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBUILD_VERSION=zephyr-v2.2.0-1790-g651b4a4e791f -DKERNEL -DTFM_PARTITION_TEST_CORE_IPC -DTFM_PSA_API -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -Itfm/install/export/tfm/include -I/projects/github/ncs/zephyr/include -Izephyr/include/generated -I/projects/github/ncs/zephyr/soc/arm/arm/mps2 -I/projects/github/ncs/modules/hal/cmsis/CMSIS/Core/Include -isystem /projects/github/ncs/zephyr/lib/libc/minimal/include -isystem /opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include -isystem /opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include-fixed -Os -imacros/projects/github/ncs/zephyr/samples/tfm_integration/tfm_ipc/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mcpu=cortex-m33+nodsp -mthumb -mabi=aapcs -march=armv8-m.main -imacros/projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-address-of-packed-member -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/projects/github/ncs/zephyr/samples/tfm_integration/tfm_ipc=CMAKE_SOURCE_DIR -fmacro-prefix-map=/projects/github/ncs/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/projects/github/ncs=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj   -c ../src/main.c
../src/main.c: In function 'psa_call':
../src/main.c:95:37: warning: passing argument 2 of 'tfm_psa_call_veneer' makes pointer from integer without a cast [-Wint-conversion]
   95 |  return tfm_psa_call_veneer(handle, (uint32_t)type,
      |                                     ^~~~~~~~~~~~~~
      |                                     |
      |                                     unsigned int
In file included from ../src/main.c:10:
tfm/install/export/tfm/include/tfm_api.h:123:70: note: expected 'const struct tfm_control_parameter_t *' but argument is of type 'unsigned int'
  123 |                                const struct tfm_control_parameter_t *ctrl_param,
      |                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
../src/main.c:97:6: warning: passing argument 4 of 'tfm_psa_call_veneer' from incompatible pointer type [-Wincompatible-pointer-types]
   97 |      (psa_invec *) &out_vecs);
      |      ^~~~~~~~~~~~~~~~~~~~~~~
      |      |
      |      psa_invec * {aka struct psa_invec *}
In file included from ../src/main.c:10:
tfm/install/export/tfm/include/tfm_api.h:125:44: note: expected 'psa_outvec *' {aka 'struct psa_outvec *'} but argument is of type 'psa_invec *' {aka 'struct psa_invec *'}
  125 |                                psa_outvec *out_vec);
      |                                ~~~~~~~~~~~~^~~~~~~

```

microbuilder commented:
> Made a couple of small suggestions for #19985 just to get it to build.
> Those are found here:
> [karl-zh#1](https://github.com/karl-zh/zephyr/pull/1)
> [zephyrproject-rtos/trusted-firmware-m#3](https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/3)

Thanks, tested on the musca_b1 and an521 build targets, and resolved the linker issue (it was due to the output path changing for the TF-M build), and works as expected now, once [zephyrproject-rtos/trusted-firmware-m#3](https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/3) is merged in.

tejlmand commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,28 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+project(NONE)
```
Why is a `project` command needed here ?
It seems as the TF-M itself is build as `ExternalProject`.

tejlmand commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2019 Linaro
```
This file should be named `<tfm-something>.cmake` instead, as it is a file being included as kind of boiler plating.

microbuilder replied:
renamed to `tfm_common.cmake`

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,72 @@
+			 ${APPLICATION_SOURCE_DIR}/build/tfm/install/outputs/AN521/tfm_s.bin
+			 ${CMAKE_BINARY_DIR}/tfm_s_signed.bin
+
+		#Sign non-secure binary image with public key
+		COMMAND ${PYTHON_EXECUTABLE} ${TFM_MCUBOOT_DIR}/scripts/imgtool.py
+		ARGS sign
+			 --layout ${PREPROCESSED_FILE}_ns.c
+			 -k ${TFM_KEY_FILE_NS}
+			 --public-key-format ${TFM_PUBLIC_KEY_FORMAT}
+			 --align 1
+			 -v ${TFM_IMAGE_VERSION_NS}
+			 ${ADD_S_IMAGE_MIN_VER}
+			 ${ADD_SECURITY_COUNTER_NS}
+			 -H 0x400
+			 ${CMAKE_BINARY_DIR}/zephyr/zephyr.bin
+			 ${CMAKE_BINARY_DIR}/zephyr_ns_signed.bin
+
+		#Create concatenated binary image from the two independently signed binary file
+		COMMAND ${PYTHON_EXECUTABLE} ${TFM_MCUBOOT_DIR}/scripts/assemble.py
+		ARGS --layout ${PREPROCESSED_FILE}_s.c
+			 -s ${CMAKE_BINARY_DIR}/tfm_s_signed.bin
+			 -n ${CMAKE_BINARY_DIR}/zephyr_ns_signed.bin
+			 -o ${CMAKE_BINARY_DIR}/tfm_sign.bin
+
+		#Copy mcuboot.bin
+		COMMAND cp ${APPLICATION_SOURCE_DIR}/build/tfm/bl2/ext/mcuboot/mcuboot.bin ${CMAKE_BINARY_DIR}
+
+		#Merge mcuboot.bin and tfm_sign.bin for QEMU
+		COMMAND ${SREC_CAT}
+		ARGS ${CMAKE_BINARY_DIR}/mcuboot.bin -Binary
+			${CMAKE_BINARY_DIR}/tfm_sign.bin -Binary
+			-offset 0x80000
+			-o ${CMAKE_BINARY_DIR}/tfm_qemu.bin -Binary
+
+		#Convert tfm_qemu.bin to .hex with an appropriate offset
+		COMMAND ${SREC_CAT}
+		ARGS ${CMAKE_BINARY_DIR}/tfm_qemu.bin -binary
+			-offset 0x10000000
+			-o ${CMAKE_BINARY_DIR}/tfm_qemu.hex -intel --line-length=44
+	)
```
This appears like something that could be useful in a more generic way, as more board could make use of TF-M.

Would it be possible to extract to common TF-M related cmake file.

microbuilder replied:
The offsets and some of the paths are going to be board specific most of the time, here. Is it worth splitting the image signing and merging process into two files in that case?

tejlmand replied:
I was more considering if there could be some reusable functions.
But the tricky part is that the board CMakeLists.txt is included before any module code.
And it is a bit specific for the extensions.cmake, although such a function could go there, and then the board could provide the needed knowledge.

I suspect some of the offsets is dts related ?

Will try to look a little into this.

tejlmand commented on boards/arm/v2m_musca_b1/CMakeLists.txt:
```diff
@@ -7,3 +7,48 @@
+	set(SREC_CAT "srec_cat")
+
+	#Create and sign for concatenated binary image should align with the TF-M BL2
+	set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+		#Create concatenated binary image from the two binary file
+		COMMAND ${PYTHON_EXECUTABLE} ${TFM_MCUBOOT_DIR}/scripts/assemble.py
+		ARGS --layout ${PREPROCESSED_FILE}.c
+			-s ${APPLICATION_SOURCE_DIR}/build/tfm/install/outputs/MUSCA_B1/tfm_s.bin
+			-n ${CMAKE_BINARY_DIR}/zephyr/zephyr.bin
+			-o ${CMAKE_BINARY_DIR}/tfm_full.bin
+
+		#Sign concatenated binary image with default public key in mcuboot folder
+		COMMAND ${PYTHON_EXECUTABLE} ${TFM_MCUBOOT_DIR}/scripts/imgtool.py
+		ARGS sign
+			--layout ${PREPROCESSED_FILE}.c
+			-k ${TFM_KEY_FILE_S}
+			--public-key-format ${TFM_PUBLIC_KEY_FORMAT}
+			--align 1
+			-v ${TFM_IMAGE_VERSION_S}
+			${ADD_SECURITY_COUNTER}
+			-H 0x400
+			${CMAKE_BINARY_DIR}/tfm_full.bin
+			${CMAKE_BINARY_DIR}/tfm_sign.bin
+
+		#srec_cat to combine images into hex for drag and drop
+		COMMAND ${SREC_CAT}
+		ARGS ${APPLICATION_SOURCE_DIR}/build/tfm/bl2/ext/mcuboot/mcuboot.bin -binary
+			-offset 0xA000000
+			${CMAKE_BINARY_DIR}/tfm_sign.bin -binary
+			-offset 0xA020000
+			-o ${CMAKE_BINARY_DIR}/tfm_zephyr.hex -Intel --line-length=44
+	)
```
This appears like something that could be useful in a more generic way, as more board could make use of TF-M.

Would it be possible to extract to common TF-M related cmake file.

tejlmand commented on samples/tfm_integration/tfm_ipc/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+
+set(QEMU_KERNEL_OPTION "-device;loader,file=${CMAKE_BINARY_DIR}/tfm_qemu.hex")
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`
```

tejlmand reviewed (changes requested):
Thanks for this, but I think we should look into a little cleanup / organization of the files.

microbuilder commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,28 @@
+
+# Set default image versions if not defined elsewhere
+if (NOT DEFINED TFM_IMAGE_VERSION_S)
+	set(TFM_IMAGE_VERSION_S 0.0.0+0)
```
This one shouldn't have a default KConfig value, in my opinion, and should be controlled and incremented by the build system, where we start at 0 if no build count file (generated during the TF-M build) is found. The override option is provided here for edge cases where you might want to continue from a previous value, and you can set the expected version as a build paramter, but generally when `TFM_IMAGE_VERSION_S` isn't already defined, it means no count file is found and we should start from 0.

microbuilder commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,28 @@
+
+# Use default signing keys if not defined elsewhere
+if (NOT DEFINED TFM_KEY_FILE_S)
+	set(TFM_KEY_FILE_S  "${TFM_MCUBOOT_DIR}/root-rsa-3072.pem")
```
Resolved in https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/773c9cc0a495afaaf133aea6f575d23249c60a12

microbuilder commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,28 @@
+	set(TFM_KEY_FILE_S  "${TFM_MCUBOOT_DIR}/root-rsa-3072.pem")
+endif()
+if (NOT DEFINED TFM_KEY_FILE_NS)
+	set(TFM_KEY_FILE_NS "${TFM_MCUBOOT_DIR}/root-rsa-3072_1.pem")
```
Resolved in https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/773c9cc0a495afaaf133aea6f575d23249c60a12

oyvindronningstad reviewed (changes requested):
Very good to see this work started.

I have some thoughts about the IPC sample. The main file is largely copied from test/suites/ipc/non_secure/ipc_ns_interface_testsuite.c in TF-M. I would really prefer that this file is used as is, from TF-M. This involves adding some more files to the build to allow calling the PSA functions by their proper name (not calling the `*_veneer()` functions directly). Also, AFAIU "IPC" in TF-M means calling the Secure APIs through a mailbox, which will need additional files added (see app/CMakeLists.txt for more info).

The most important thing IMO is allowing the Zephyr user to call PSA APIs as they would if they developed on TF-M directly, even if we don't support both IPC and library calls.

See this if you haven't before: https://ci.trustedfirmware.org/job/tf-m-build-test-nightly/lastSuccessfulBuild/artifact/build-docs/tf-m_documents/install/doc/user_guide/html/docs/user_guides/tfm_build_instruction.html#configurations

microbuilder commented:
> Very good to see this work started.
> 
> I have some thoughts about the IPC sample. The main file is largely copied from test/suites/ipc/non_secure/ipc_ns_interface_testsuite.c in TF-M. I would really prefer that this file is used as is, from TF-M. This involves adding some more files to the build to allow calling the PSA functions by their proper name (not calling the `*_veneer()` functions directly). Also, AFAIU "IPC" in TF-M means calling the Secure APIs through a mailbox, which will need additional files added (see app/CMakeLists.txt for more info).
> 
> The most important thing IMO is allowing the Zephyr user to call PSA APIs as they would if they developed on TF-M directly, even if we don't support both IPC and library calls.
> 
> See this if you haven't before: https://ci.trustedfirmware.org/job/tf-m-build-test-nightly/lastSuccessfulBuild/artifact/build-docs/tf-m_documents/install/doc/user_guide/html/docs/user_guides/tfm_build_instruction.html#configurations

I agree this could be done differently, and I have two other demos that I need to get in, one using library mode, and the other to run the PSA FF API, etc., tests as a sample app, but this might need to happen after the 2.3 merge window closes Friday.

oyvindronningstad commented:
> @oyvindronningstad I agree this could be done differently, and I have two other demos that I need to get in, one using library mode, and the other to run the PSA FF API, etc., tests as a sample app, but this might need to happen after the 2.3 merge window closes Friday.

Sounds good. The one thing I want to get in before the merge window is allowing the PSA APIs to be used by their proper names, so that once the PSA APIs are added, it doesn't break things for the user.

tejlmand commented:
@microbuilder Placed some additional code, just to provide a starting point, without changing current design.
https://github.com/karl-zh/zephyr/pull/2
https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/6

Basically, it move the `ExternalProject_Add()` to the TFM Zephyr module, and then calls `trusted_firmware_build` instead.

Now, the question, which I think @oyvindronningstad is also addressing, is how do we want to integrate TF-M into Zephyr in general.

Cause I don't assume we want to add `trusted_firmware_build` throughout Zephyr samples.
Also, there could be downstream users with a different solution, or not wanting to clone TF-M, and such a function will break the build, if it is missing.

Do you have some thoughts regarding the future integration that  you can share ?

tejlmand commented:
which actually makes me notice there is no documentation accompanying this PR.

It should be documented how users can enable and use TF-M.

microbuilder commented:
> which actually makes me notice there is no documentation accompanying this PR.
> 
> It should be documented how users can enable and use TF-M.

There is some documentation at the board level, but no high-level TF-M documentation, I agree. I'll add something to get started in with the next push, once I'm done some further cleanup with the build system.

carlescufi commented on samples/tfm_integration/tfm_ipc/README.rst:
```diff
@@ -0,0 +1,103 @@
+.. _tfm_ipc:
+
+TF-M IPC
```
Please add a small introduction to TF-M and links to it.

microbuilder replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/e5183e90d2655ed42bc9d7cf55cbf9529a0d777f

carlescufi commented:
@microbuilder this needs to be moved out of Draft ASAP if it is to be merged in time for 2.3.

microbuilder commented:
> @microbuilder this needs to be moved out of Draft ASAP if it is to be merged in time for 2.3.

@carlescufi Making some final changes to it tonight, and hope to flip the switch tomorrow morning for review. Sorry for the delay.

microbuilder commented on samples/tfm_integration/CMakeLists.inc:
```diff
@@ -0,0 +1,28 @@
+project(NONE)
+
+# Set TF-M folders
+set(TFM_REMOTE_DIR "$ENV{ZEPHYR_BASE}/../modules/tee/tfm")
```
This has been refactored, and this file no longer exists.

microbuilder commented:
> > @oyvindronningstad I agree this could be done differently, and I have two other demos that I need to get in, one using library mode, and the other to run the PSA FF API, etc., tests as a sample app, but this might need to happen after the 2.3 merge window closes Friday.
> 
> Sounds good. The one thing I want to get in before the merge window is allowing the PSA APIs to be used by their proper names, so that once the PSA APIs are added, it doesn't break things for the user.

@oyvindronningstad This commit should resolve the issue you brought up, and PSA API functions can now be called directly from the sample app: https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/ce2726d1b603157b4d1c62c61ec019613ce11424

microbuilder committed:
samples: tfm_integration: Add PSA level 1 sample app

This commit adds a `tfm_level_1` samples app that shows how
to use the PSA APIs in IPC mode in a real-world example.

It makes use of the crypto, initial attestation and secure
storage modules.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>

microbuilder commented:
@oyvindronningstad I also added a more complex 'real world' example making use of the attestation, crypto and secure storage modules here, accessing the PSA APIs with the normal names in IPC mode of course: https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/4e2173e3269f70559ae6e9ea79bbd3ebe5aaa582

tejlmand commented on samples/tfm_integration/psa_level_1/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+# Set "tfm" as a dependency so that it's built before the app
+add_dependencies(app tfm)
+
+# Source files in this sample
+target_sources(app PRIVATE src/main.c)
+target_sources(app PRIVATE src/psa_attestation.c)
+target_sources(app PRIVATE src/psa_crypto.c)
+target_sources(app PRIVATE src/shell.c)
+target_sources(app PRIVATE src/tfm_ipc.c)
+target_sources(app PRIVATE src/util_app_cfg.c)
+target_sources(app PRIVATE src/util_app_log.c)
+target_sources(app PRIVATE src/util_sformat.c)
+
+# Include "/install/export/tfm/include" when building
+target_include_directories(app PRIVATE ${TFM_INCLUDE_PATH})
+# Account for "psa/crypto_compat.h" not being in installs until fixed
+target_include_directories(app PRIVATE ${ZEPHYR_TFM_MODULE_DIR}/trusted-firmware-m/interface/include)
+
+# Include TF-M secure service source files
+set(TFM_ROOT_PATH ${CMAKE_BINARY_DIR}/tfm/install/export/tfm/src/)
+target_sources(app PRIVATE ${TFM_ROOT_PATH}/tfm_crypto_ipc_api.c)
+target_sources(app PRIVATE ${TFM_ROOT_PATH}/tfm_initial_attestation_ipc_api.c)
+target_sources(app PRIVATE ${TFM_ROOT_PATH}/tfm_its_ipc_api.c)
+target_sources(app PRIVATE ${TFM_ROOT_PATH}/tfm_ns_interface.c)
+target_sources(app PRIVATE ${TFM_ROOT_PATH}/tfm_platform_ipc_api.c)
+target_sources(app PRIVATE ${TFM_ROOT_PATH}/tfm_psa_ns_api.c)
+target_sources(app PRIVATE ${TFM_ROOT_PATH}/tfm_sst_ipc_api.c)
```
To be update similar to this: https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/eb3c635377dda57e7f4c2aca83574465d4f5b595
and:
https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/6/commits/55f5ece809871bfd9756754da63faa557eb93d33

microbuilder replied:
See: https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/532792f8490244e54a274b1834cb919401c62fe6

Verified with `$ west build -p -b mps2_an521_nonsecure zephyr/samples/tfm_integration/psa_level_1/ -t run`

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -1,9 +1,85 @@
+	endif()
+
+	set(PREPROCESSED_FILE "${CMAKE_BINARY_DIR}/tfm/image_macros_preprocessed")
+	set(TFM_MCUBOOT_DIR "$ENV{ZEPHYR_BASE}/../modules/tee/tfm/trusted-firmware-m/bl2/ext/mcuboot")
```
$ENV{ZEPHYR_BASE} is not mandatory, so this might fail.

Use ${ZEPHYR_BASE} instead.

microbuilder replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/62b2df85cbd669ba451693bcb731137af4140e90

microbuilder commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -1,9 +1,85 @@
+		COMMAND cp ${CMAKE_BINARY_DIR}/tfm/bl2/ext/mcuboot/mcuboot.bin ${CMAKE_BINARY_DIR}
+
+		#Merge mcuboot.bin and tfm_sign.bin for QEMU
+		COMMAND ${SREC_CAT}
```
We do document this as a requirement: https://github.com/zephyrproject-rtos/zephyr/pull/19985/files#diff-f56b3c0c93cae2c1f9198d0c24dd854eR57 ... but it should use `find_program`, yes.

microbuilder replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/62b2df85cbd669ba451693bcb731137af4140e90

ioannisg commented on modules/Kconfig.tfm:
```diff
@@ -0,0 +1,25 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BUILD_WITH_TFM
+	bool "Build with TF-M for Secure Execution"
```
This option requires a help text. 
We need to explain what it means "to build with TF-M for Secure Execution":

For instance: shall this mean "Build a NON-SECURE Zephyr execution environment with TF-M for Secure Execution"?
If so I would add a "depends on TRUSTED_EXECUTION_NONSECURE"
and then this would mean that:
in order to build a non-secure Zephyr firmware with TF-M I need to
- select an "_ns" Zephyr board target (which selects TRUSTED_EXECUTION_NONSECURE"
- enable CONFIG_BUILD_WITH_TFM"
In order to build a

microbuilder replied:
Is this sufficiently clear, in your opinion?

```
config BUILD_WITH_TFM
	bool "Build with TF-M for Secure Execution"
	depends on TRUSTED_EXECUTION_NONSECURE
	help
	  This will cause the Zephyr application to be placed in the
	  non-secure processing environment, and a TF-M image will be
	  generated for the secure processing environment. Both
	  images, as well as the veneer object file that links them, will
	  be generated during the normal Zephyr build process.

	  When building with CONFIG_BUILD_WITH_TFM enabled, you must use
	  the "*_nonsecure" board variant (ex. "mps2_an521_nonsecure").
```

ioannisg replied:
Yeap, the main points are covered, let me work on an iteration (will post it here)

ioannisg replied:
```
config BUILD_WITH_TFM
	bool "Build with TF-M as the Secure Execution Environment"
	depends on TRUSTED_EXECUTION_NONSECURE
	help
	  When enabled, this option instructs the Zephyr build process to
          additionaly generate a TF-M image for the Secure Execution
          environment, along with the Zephyr image. The Zephyr image
          itself is to be executed in the Non-Secure Processing Environment.
          The required dependency on TRUSTED_EXECUTION_NONSECURE
          ensures that the Zephyr image is built as a Non-Secure image. Both
	  TF-M and Zephyr images, as well as the veneer object file that links
          them, are generated during the normal Zephyr build process.

          Note:
	    Building with the "*_nonsecure" BOARD variant (e.g.
            "mps2_an521_nonsecure") ensures that
            CONFIG_TRUSTED_EXECUTION_NONSECURE ie enabled.
```

ioannisg replied:
Maybe this one @microbuilder @tejlmand ? LAtest version is more explanatory around the dependency on TRUSTED_EXECUTION_NONSECURE.

microbuilder replied:
Updated with this, thanks for the suggestion. Agree it's clearer.

ioannisg commented on samples/tfm_integration/psa_level_1/Kconfig:
```diff
@@ -0,0 +1,23 @@
+
+config PSA_SHELL
+	bool "Enable the 'psa' shell command"
+	default n
```
default n is not needed

Commit:
trusted-firmware-m: Add initial plumbing for TFM

This commits adds a config option to build a PSA trusted-firmware-m
(TF-M) application image for the secure processing environment, and
configures the Zephyr application image for the non-secure processing
environment. The secure and non-secure environment images will be
linked together via the veneer function table that is produced as an
artifact of the TF-M build process.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

Commit:
boards: arm: mps2_an521: Add TFM support

This commit adds support for TF-M to the MPS2 AN521.

When the CONFIG_BUILD_WITH_TFM flag is set, a secure and
non-secure processing environment image pair will be
generated, with the Zephyr application image running on
the non-secure side.

The secure and non-secure binary images will be signed
for use with the BL2 secure bootloader.

An additional .hex file is also generated to enable
running QEMU with the AN521 binaries, `tfm_qemu.hex`,
which can be executed with the `-t run` option with
west,  or `run` with ninja or make.

When configured for use with TF-M, the
`mps2_an521_nonsecure` board definition should be used.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

Commit:
boards: arm: musca_b1: Add TFM support

This commit adds support for TF-M to the MUSCA B1.

When the CONFIG_BUILD_WITH_TFM flag is set, a secure and
non-secure processing environment image pair will be
generated, with the Zephy application image running on
the non-secure side.

The secure and non-secure binary images will be signed
for use with the BL2 secure bootloader.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

Commit:
samples: tfm_integration: Add TFM IPC sample app

This commit adds a minimal sample application showing how
TF-M can be used in IPC mode, with Zephyr providing the
non-secure processing environment image, and linking against
the PSA APIs implemented in TF-M.

This commit also include work from the commit quoted below,
which is included here for attribution purposes:

    tfm_ipc: cmake: now using tfm_ipc library created by TF-M module

    This commit removes the need for local knowledge of the TF-M
    repository structure and files.

    Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -1,9 +1,85 @@
+			 -o ${CMAKE_BINARY_DIR}/tfm_sign.bin
+
+		#Copy mcuboot.bin
+		COMMAND cp ${CMAKE_BINARY_DIR}/tfm/bl2/ext/mcuboot/mcuboot.bin ${CMAKE_BINARY_DIR}
```
This is not portable to Windows.
Please use: `${CMAKE_COMMAND} -E copy`
https://cmake.org/cmake/help/latest/manual/cmake.1.html#run-a-command-line-tool

microbuilder replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/e078c697b48f5b0d3ec474b05eaa4d80aee7dd30

ioannisg commented on soc/arm/arm/mps2/soc.c:
```diff
@@ -35,6 +35,7 @@ FPGAIO_INIT(2);
 
 #define NON_SECURE_FLASH_ADDRESS	(0x100000)
 #define NON_SECURE_FLASH_OFFSET	(0x10000000)
+#define BL2_HEADER_SIZE		(0x400)
```
I think all these 3 macros need to be documented - e.g. are they specific to Tf-M? E.g. the BL2 header size being 0x400.

microbuilder replied:
I'm not intimitely familiar with this FPGA board and image, which is a bit of an oddity with some unusual design decisions, but this code reserves space for BL2 on the second (non-secure) CPU whether the bootloader is used or not.

This is used in the `ipm_mhu_dual_core` sample, not for TF-M, which currently only makes use of a single core in existing examples for Zephyr. 

I'll document them in this PR as best I understand them, though.

microbuilder replied:
```
/* The starting address for the application image on the secondary
 * (non-secure) cpu.
 */
#define NON_SECURE_FLASH_ADDRESS	(0x100000)

/* The offset for the secondary (non-secure) cpu image, which is used
 * to determine the location of the reset vector at startuo.
 */
#define NON_SECURE_FLASH_OFFSET	        (0x10000000)

/* Space reserved for TF-M's secure bootloader on the non-secure cpu.
 * This space is reserved whether BL2 is used or not.
 */
#define BL2_HEADER_SIZE                 (0x400)
```
?

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -1,9 +1,85 @@
+	set(TFM_PUBLIC_KEY_FORMAT "full")
+
+	# Set srec_cat binary name
+	find_program(SREC_CAT srec_cat)
```
Not enough.
You need to safe guard, and maybe even inform the users that there are missing dependencies:
```
**[INFO]** Image version number set to 0.0.0+1
**[INFO]** Image version number set to 0.0.0+2
/bin/sh: 1: SREC_CAT-NOTFOUND: not found
ninja: build stopped: subcommand failed.
```

microbuilder replied:
```
# Set srec_cat binary name
find_program(SREC_CAT srec_cat)
if(NOT SREC_CAT)
    message(FATAL_ERROR "'srec_cat' not found. Please install it, or add it to $PATH.")
endif()
```
Is this check sufficient, in your opinion?

tejlmand replied:
will not work:
>  If nothing is found, the result will be `<VAR>-NOTFOUND`
https://cmake.org/cmake/help/v3.17/command/find_program.html

You need to check as:
```
# Set srec_cat binary name
find_program(SREC_CAT srec_cat)
if(${SREC_CAT} STREQUAL SREC_CAT-NOTFOUND)
    message(FATAL_ERROR "'srec_cat' not found. Please install it, or add it to $PATH.")
endif()
```

ioannisg commented on boards/arm/mps2_an521/board.cmake:
```diff
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 set(EMU_PLATFORM qemu)
+set(TFMBOARD "AN521")
```
Can you rename to TFM_BOARD, looks nicer.
Alternatively you can use TFM's terminology : TARGET_PLATFORM

microbuilder replied:
Maybe `TFM_TARGET_PLATFORM`, though large, is the clearest?

ioannisg replied:
Yes, that's good.

ioannisg commented on samples/tfm_integration/tfm_ipc/prj.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_BUILD_WITH_TFM=y
```
Could you add some of the options that are added in level_1 prj.conf?
E.g. configurations for HEAP, thread options, etc. I would think they would be useful

microbuilder replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/19985/commits/e078c697b48f5b0d3ec474b05eaa4d80aee7dd30

ioannisg commented:
Just a gentle reminder @microbuilder (in case you 've forgotten) there are some commit messages to be filled in.

microbuilder commented:
> Just a gentle reminder @microbuilder (in case you 've forgotten) there are some commit messages to be filled in.

@ioannisg Was just fixing those, done now.

carlescufi commented:
Tagging as DNM because of the missing SHA in `west.yml`. Otherwise this looks good to go to me.

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -1,9 +1,88 @@
+
+	# Set srec_cat binary name
+	find_program(SREC_CAT srec_cat)
+	if(NOT SREC_CAT)
```
```suggestion
 if(${SREC_CAT} STREQUAL SREC_CAT-NOTFOUND)
```

tejlmand commented on boards/arm/v2m_musca_b1/CMakeLists.txt:
```diff
@@ -1,9 +1,65 @@
+
+	# Set srec_cat binary name
+	find_program(SREC_CAT srec_cat)
+	if(NOT SREC_CAT)
```
```suggestion
  if(${SREC_CAT} STREQUAL SREC_CAT-NOTFOUND)
```

tejlmand reviewed (changes requested):
almost there.

tejlmand reviewed (approved):
Approved.
Waiting for sha instead of pull/6/head.

ioannisg commented on soc/arm/arm/mps2/soc.c:
```diff
@@ -33,8 +33,20 @@ FPGAIO_INIT(2);
-#define NON_SECURE_FLASH_OFFSET	(0x10000000)
+
+/* The offset for the secondary (non-secure) cpu image, which is used
+ * to determine the location of the reset vector at startuo.
```
startup*

ioannisg commented on soc/arm/arm/mps2/soc.c:
```diff
@@ -33,8 +33,20 @@ FPGAIO_INIT(2);
 #define SSE_200_SYSTEM_CTRL_CPU_WAIT	(SSE_200_SYSTEM_CTRL_S_BASE + 0x118)
 #define SSE_200_CPU_ID_UNIT_BASE	(0x5001F000UL)
 
+/* The starting address for the application image on the secondary
```
@microbuilder @galak i still cannot  imeediately understand which of the 2 is correct:
- there is a second CPU on AN521, which just does not have TrustZone and is by default non-secure
- you refer to the non-secure image address offsets of the CPU we build Zephyr on.

ioannisg replied:
here you are using "non secure cpu", later, "non secure cpu image"

microbuilder replied:
It's the base address that the application image running on the secondary mcu starts at. It seems clear to me that an image runs on an mcu, and it starts at a specific address??

If you want to propose something specific, I'm happy to change it.

ioannisg replied:
so i understand that there are 3 images you are talking about:
- tf-m runs as the SPE on primary CPU
- Zephyr runs as the NSPE on primary CPU
- and there is another firmware running on a secondary CPU

correct @microbuilder ?

microbuilder replied:
```
/* The base address tha the application image will start at on the secondary
 * (non-TrustZone) Cortex-M33 mcu.
 */
#define NON_SECURE_FLASH_ADDRESS	(0x100000)

/* The memory map offset for the application image, which is used
 * to determine the location of the reset vector at startup.
 */
#define NON_SECURE_FLASH_OFFSET	        (0x10000000)

/* Space reserved for TF-M's secure bootloader on the secondary mcu.
 * This space is reserved whether BL2 is used or not.
 */
#define BL2_HEADER_SIZE                 (0x400)
```

We can clarify this with ARM or dig into the DS to change this later, but seems silly to hold up the TF-M merge tonight for comments on fields that aren't used, even if I agree clear documentation is important?

ioannisg replied:
@microbuilder we are not at all holding up the merge - just want things to look nice.


So, since we now clarified that this is related to an image of a second Cortex-M33 on AN521, that is non-TrustZone enabled, I suggest you rename the macros, and remove the NONSECURE prefix, because this made me think the offsets are related to the non-secure NSPE

microbuilder replied:
>     * and there is another firmware running on a secondary CPU, which happens to be non-secure

The first two are right, yes.

Nothing is running on the secondary CPU, though. This code for the second CPU is never used by the TF-M demos, but seems to have been added for the following demos:

```
samples/subsys/ipc/ipm_mhu_dual_core/src/main.c:65:          wakeup_cpu1();
samples/subsys/ipc/openamp/src/main.c:294:   wakeup_cpu1();
```

ioannisg replied:
> Nothing is running on the secondary CPU, though. This code for the second CPU is never used by the TF-M demos, but seems to have been added for the following demos:

Ok thanks a lot for clarifying. In any case, I believe you should remove the NONSECURE prefix from these local macros, because it is confusing; I thought they refer to the Zephyr NSPE.
The documentation you added is ok.

microbuilder replied:
Renamed values to `CPU1_*` and tested the two IPC samples, and seems OK, and agree NONSECURE is confusing.

carlescufi commented:
@ioannisg take another look please, commits have been rebased into a logical set

ioannisg commented:
@microbuilder thanks for taking over this PR and for bringing this to merge-able state
@tejlmand @oyvindronningstad thanks for the reviews!
